### PR TITLE
ENG-15956: Major rewrite of the JUnit auto-filer.

### PR DESCRIPTION
**Pull request by johnTcrawford**
A completed and fleshed out version of my (J. Crawford's) Innovation
Week project.

### CODE REVIEW

File: tools/jenkins/jenkinsbot.py [hunk 1]
```diff
@@ -736,10 +737,695 @@ def query_and_response(self, query, params, channels, filename, vertical=False,
...
+               find a related, open Jira ticket; typically the first one is the Test Suite in which the failed
+               test exists, and the second is the name of the test itself.
+        :param labels: The Labels to list in the Jira ticket.
+        :param jira: A JIRA access object, used to create a Jira ticket; if not
+               specified, the user and passwd will be used to create one.
+        :param user: The Jira Username used to access Jira.
+        :param passwd: The Jira Password for that User.
+        :param project: The Jira Project in which the Jira ticket should be created.
+        """
+        logging.debug('In find_jira_bug_ticket:')
```
Comment (kaikaizi) [inline tools/jenkins/jenkinsbot.py]: Could this method call `find_jira_bug_tickets` and return the first one?

File: tools/jenkins/jenkinsbot.py [hunk 2]
```diff
@@ -736,10 +737,695 @@ def query_and_response(self, query, params, channels, filename, vertical=False,
...
+
+
+    def is_number(self, s):
+        try:
+            return float(s)
+        except ValueError:
+            return False
+
+
+    def get_jira_components(self, component, jira, project=JIRA_PROJECT):
```
Comment (kaikaizi) [inline tools/jenkins/jenkinsbot.py]: It seems that `get_first_jira_component` is a more accurate name.

File: tools/jenkins/jenkinsbot.py [hunk 3]
```diff
@@ -736,10 +737,695 @@ def query_and_response(self, query, params, channels, filename, vertical=False,
...
+        jira_component = 'Core'  # default, since Components is a required field
```
Comment (kaikaizi) [inline tools/jenkins/jenkinsbot.py]: It could be better to make this an optional argument.

File: tools/jenkins/jenkinsbot.py [hunk 4]
```diff
@@ -736,10 +737,695 @@ def query_and_response(self, query, params, channels, filename, vertical=False,
...
+                jira_component = {
+                    'name': c.name,
+                    'id': c.id
+                }
+                break
+        return [jira_component]
+
+
+    def get_jira_versions(self, version, jira, project=JIRA_PROJECT):
+        jira_version = 'Backlog'  # default, since Versions is a required field
```
Comment (kaikaizi) [inline tools/jenkins/jenkinsbot.py]: It could be better to make this an optional argument.

File: tools/jenkins/jenkinsbot.py (hunk 1 above)
Comment (johnTcrawford) [reply tools/jenkins/jenkinsbot.py]: That's what it does. (And logs a Warning message if there is more than one.)

File: tools/jenkins/jenkinsbot.py (hunk 2 above)
Comment (johnTcrawford) [reply tools/jenkins/jenkinsbot.py]: Yes, I'll fix that.

File: tools/jenkins/jenkinsbot.py (hunk 3 above)
Comment (johnTcrawford) [reply tools/jenkins/jenkinsbot.py]: Done (and similarly in get_jira_version_list).

File: tools/jenkins/jenkinsbot.py (hunk 4 above)
Comment (johnTcrawford) [reply tools/jenkins/jenkinsbot.py]: Done (even before I saw this comment - see above).

**Pull request merged**